### ISSUE TITLE
seq2seq bidirection with attention throws code error

### ISSUE THREAD

**Issue by hsgodhia** [facebookresearch/ParlAI#542]
Hey

When using the following command on pytorch 0.3 (py3.6)
```
harshals@tesla:~/ParlAI$ python3.6 examples/train_model.py -e 50 -m seq2seq -mf godz5 -t opensubtitles -dt train:stream -hist 100 -bs 1 -tr 48 --dict-maxexs=10000000 --gpu 2 --batch-sort false -hs 512 -esz 300 -nl 2 -emb glove -lm both -bi true -att dot

```

I get the following error stack

```
[loading fbdialog data:/home/harshals/ParlAI/data/OpenSubtitles/train.txt]
Traceback (most recent call last):
  File "examples/train_model.py", line 237, in <module>
    TrainLoop(setup_args()).train()
  File "examples/train_model.py", line 201, in train
    world.parley()
  File "/home/harshals/ParlAI/parlai/core/worlds.py", line 291, in parley
    acts[1] = agents[1].act()
  File "/home/harshals/ParlAI/parlai/agents/seq2seq/seq2seq.py", line 594, in act
    return self.batch_act([self.observation])[0]
  File "/home/harshals/ParlAI/parlai/agents/seq2seq/seq2seq.py", line 541, in batch_act
    _, _, loss = self.predict(xs, ys, lm=True)
  File "/home/harshals/ParlAI/parlai/agents/seq2seq/seq2seq.py", line 390, in predict
    predictions, scores, _ = self.model(xs, ys)
  File "/usr/local/lib/python3.6/site-packages/torch/nn/modules/module.py", line 325, in __call__
    result = self.forward(*input, **kwargs)
  File "/home/harshals/ParlAI/parlai/agents/seq2seq/modules.py", line 80, in forward
    preds, score, _h = self.decoder(xi, hidden, enc_out, attn_mask)
  File "/usr/local/lib/python3.6/site-packages/torch/nn/modules/module.py", line 325, in __call__
    result = self.forward(*input, **kwargs)
  File "/home/harshals/ParlAI/parlai/agents/seq2seq/modules.py", line 224, in forward
    xes = self.attention(xes, hidden, encoder_output, attn_mask)
  File "/usr/local/lib/python3.6/site-packages/torch/nn/modules/module.py", line 325, in __call__
    result = self.forward(*input, **kwargs)
  File "/home/harshals/ParlAI/parlai/agents/seq2seq/modules.py", line 486, in forward
    torch.bmm(hid, enc_out.transpose(1, 2)).squeeze(1))
RuntimeError: invalid argument 6: wrong matrix size at /pytorch/torch/lib/THC/generic/THCTensorMathBlas.cu:453
```

- The above error comes with all 4 types of -att
- In above dec, enc do not share weights (default -dec same)
- I basically suspect that when the encoder type is a bi-directional LSTM and you combine that with attention of any kind it throws this error
- I noticed that the paper referenced for attention in ParlAI (1508.04025) doesn't consider a bi-di LSTM as opposed to Bahdanau et al

When I change the command to remove attention but keep bi direction, the below command works

```
harshals@tesla:~/ParlAI$ python3.6 examples/train_model.py -e 50 -m seq2seq -mf godz5 -t opensubtitles -dt train:stream -hist 100 -bs 1 -tr 48 --dict-maxexs=10000000 --gpu 2 --batch-sort false -hs 512 -esz 300 -nl 2 -emb glove -lm both -bi true
```

**Comment by alexholdenmiller** [comment]
Thanks for the heads up! Can you confirm which Bahdanau et al paper you're referring to?

In general--welcome to any suggestions on the best approaches for passing bidirectional state into a single-directional encoder. You can observe the current approach is just taking the element-wise max between the two directions (https://github.com/facebookresearch/ParlAI/blob/master/parlai/agents/seq2seq/modules.py#L170), but this is little more than a placeholder.

**Comment by hsgodhia** [comment]
- I see so current implementation is using the BiLSTM-max. Bahdanau Learning to Align and Translate https://arxiv.org/abs/1409.0473 They use a concatenation of fwd and bwd. 

- What about the error stack that comes, can we not use a bidirection encoder with attention ParlAI options ( -bi true -att dot) ?

**Comment by alexholdenmiller** [comment]
Thanks, that's the paper I thought you meant but just wanted to make sure. Working on fixing the stack error!

**Comment by alexholdenmiller** [comment]
Turns out we already supported attention over concatenated output states, but had lost the parameter init in a refactor. Fixed in #560. For `dot` attention, I duplicate the current hidden state to get the dimensionality right.

The `max` I mentioned before referring to the final hidden state of the decoder, which is used to initialize the hidden state of the encoder, but that initialization could probably be improved and has not been extensively evaluated.
